### PR TITLE
[CUDA] Fix linspace integer dtype precision mismatch with CPU

### LINKED ISSUE #181807: torch.linspace with integer dtype produces inconsistent results on CPU and CUDA
## Summary

`torch.linspace` produces inconsistent integer results between CPU and CUDA when `dtype=torch.int64`.

For the same `start`, `end`, `steps`, and `dtype`, CPU and CUDA generate different integer sequences.

## Reproduction

```python
import torch

cpu = torch.linspace(3.7, -3, 10, dtype=torch.int64, device="cpu")
gpu = torch.linspace(3.7, -3, 10, dtype=torch.int64, device="cuda").cpu()

print("cpu:", cpu.tolist())
print("gpu:", gpu.tolist())
print("differing indices:", (cpu != gpu).nonzero(as_tuple=True)[0].tolist())




### Versions

cpu: [3, 2, 1, 1, 0, 0, -1, -1, -2, -3]
gpu: [3, 2, 1, 0, 0, 0, 0, -1, -2, -3]
differing indices: [3, 6]

Expected behavior

CPU and CUDA should produce the same integer tensor for the same torch.linspace arguments.

Why this seems like a bug

This does not appear to be a normal tolerance issue, because the requested output dtype is torch.int64.

The discrepancy occurs in the integer values themselves:

index 3: CPU = 1,  CUDA = 0
index 6: CPU = -1, CUDA = 0

This suggests that CPU and CUDA use different rounding or casting behavior when generating integer linspace values. Since torch.linspace is a deterministic tensor construction operator, the result should ideally be backend-consistent.

PyTorch 2.11.0+cu128, CUDA 12.8

cc @ptrblck @msaroufim @eqy @jerryzh168 @tinglvv @nWEIdia @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @gchanan @mruberry

### PR DESCRIPTION
**Pull request by gaurav-redhat**
The CUDA linspace kernel used float for the step when the output dtype is integral, while the CPU kernel uses double. That can make CPU and CUDA disagree on int64 linspace for the same start, end, and steps.

Use double for the integral CUDA path to match RangeFactoriesKernel.cpp.

Fixes https://github.com/pytorch/pytorch/issues/181807

### CODE REVIEW

**Comment by gaurav-redhat** [comment]
Apologies for the noise — a force-push accidentally triggered auto-assignment of many unrelated code owners and CI labels. The only files changed are `aten/src/ATen/native/cuda/RangeFactories.cu` and `test/test_tensor_creation_ops.py`.

@pytorchbot label "release notes: cuda"

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by ezyang** [comment]
The change here isn't free. You'll be using the fp64 rather than fp32 hardware, and there's quite a bit less of it. It probably doesn't matter here, but I'm just a little bit worried. (cc @ngimel)

**Comment by gaurav-redhat** [comment]
Thanks @ezyang. This only applies to integer dtypes (float/double paths unchanged), and the kernel is memory-bound — just one int write per thread — so the fp64 step arithmetic shouldn't be measurable. Same approach the CPU path uses in RangeFactoriesKernel.cpp.

**Comment by ngimel** [comment]
It actually matters for GB300

**Review by ezyang — CHANGES_REQUESTED** [review]
Needs GB300 bench across variety of sizes showing no impact

**Comment by jeffdaily** [comment]
_This analysis was AI-assisted._

The `double` fix is correct and minimal: CPU already computes `step` in `double`, so this makes CUDA bit-identical with no behavior change. The one cost is that `step` is used per-element (`start + step * ind`), so this is a per-element `double` multiply -- which is what @ezyang's GB300 request is probing, since fp64 is rate-limited on Blackwell (and on ROCm RDNA; CDNA/MI is full-rate so unaffected). Two alternatives if we want to avoid that:

**1. Exact integer math (`__int128`).** Compute `start + (__int128(end - start) * ind) / (steps - 1)` -- integer-ALU, fast on every arch, exact, no fp64.
- Pro: eliminates the fp64 concern entirely.
- Con: integer truncation differs from `double` truncation by up to ±1 (`floor((end-start)*i/(steps-1))` vs `floor(((end-start)/(steps-1))*i)`), so CUDA-only would introduce *new* CPU/CUDA mismatches. Keeping parity means converting the CPU kernel too, which shifts existing int-`linspace` results (a BC change). More correct, but heavier and riskier than a parity fix warrants.

**2. Conditional precision.** Keep `float` when the range fits in float's exact-integer range (< 2^24, the common case), use `double` only for larger ranges that need it.
- Pro: the common path stays `float`, so there is no per-element fp64 and no regression to benchmark -- this sidesteps the GB300 ask. Large ranges still match CPU.
- Con: two kernel instantiations (templated step type) plus a host-side threshold; more code for a non-hot-path op.

My take: for a setup op that's essentially never in a hot loop, the `double` fix as-is is the pragmatic choice and the benchmark will almost certainly show no impact. If we'd rather not gate on a GB300 run at all, option 2 keeps the fast path in `float` and makes "no perf impact" true by construction.
